### PR TITLE
Support complex attributes in the Event API

### LINKED ISSUE #4199: Event API should document whether it supports complex attribute types
We restrict attribute types to primitives and array or primitives and make an exception for `LogRecord`.

https://github.com/open-telemetry/opentelemetry-specification/blob/e9a2b004ccf547871c5208aa9fff3797a84549e8/specification/common/README.md?plain=1#L76-L80

See https://github.com/open-telemetry/opentelemetry-specification/pull/3858 for the context.

Semconv event definition however limits events attributes and does not allow them to have complex types.

> It MAY have [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.35.0/specification/common#attribute)
  attributes that provide additional context about the event.

https://github.com/open-telemetry/semantic-conventions/blob/11f5d37d1624c2392ea97f40e76c1b3c7cd30124/docs/general/events.md?plain=1#L61

We should explicitly call out supported attribute types for events in the spec so that the restriction applies to all events and not just those defined in semantic conventions (or decide that events support complex types).

### PR DESCRIPTION
**Pull request by trask**
Fixes #4199

Allow complex attributes on Events.

Based on discussion in https://github.com/open-telemetry/semantic-conventions/issues/1651 and in today's Specification SIG meeting ([recording](https://zoom.us/rec/share/tMUYns0fQrFQjvQJgl3Vgttu8az2z4-40WSknkDHriMh0zjHKDI1SELJCOj9nWmz.lEpooohZ6y44t8bB)).

The biggest struggle currently for launching the Event API is the "two property bags" problem, also known as "do I put my event key/value pairs into event attributes or into the event body?"

The Log SIG has been trying to figure out a satisfactory answer to this for a month.

The proposal from @lmolkova is to put everything into event attributes, eliminating one of these "property bags" and answering the question of where to put event key/value pairs.

The event body could still potentially be used for dumping unstructured (or possibly very large) data.

To make this work, we need complex attributes on Events.

This would allow for example "locally scoped" attributes on an event to be defined as a complex attribute in the semantic conventions, e.g. an event named `browser.page_navigation_timing` could define a complex attribute in the semantic conventions named `browser.page_navigation_timing` (could be the same as the event name or different), which would contain it's "locally scoped" event attributes underneath it, e.g.

```
{
  "user.name": ...,
  ...,
  "browser.page_navigation_timing": {
    "name": ...,
    "fetchStart": ...,
    "unloadEventStart": ...,
    "unloadEventEnd": ...,
    ...
  },
  ...
}

### CODE REVIEW

**Review by lmolkova — APPROVED** [review]

**Review by pellared — APPROVED** [review]
A prototype in OTel Go: https://github.com/open-telemetry/opentelemetry-go/pull/6017

_The Go SIG strongly prefers: https://github.com/open-telemetry/opentelemetry-go/pull/6018. However, it is also using supporting complex log attributes._

**Comment by austinlparker** [comment]
Quick question/clarification - since this strikes me as extending the common attribute definition (https://opentelemetry.io/docs/specs/otel/common/), are we limiting the depth of complex attributes?

**Comment by reyang** [comment]
We have limits on the number of attributes https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#readablelogrecord. This PR would give people a workaround if they just add everything to a single complex attribute - which is concerning from reliability/performance/security perspective.

**Comment by pellared** [comment]
>  This PR would give people a workaround if they just add everything to a single complex attribute - which is concerning from reliability/performance/security perspective.

The attribute limiting issue is already present in the existing Logs (Bridge) API. What is more there are no limits for the Body; see:
- https://github.com/open-telemetry/opentelemetry-specification/issues/4006

**Comment by trask** [comment]
We discussed in the Log SIG yesterday and decided that these both need to be addressed before stabilizing Events:

* Length limits for complex attribute bodies
* What does "number of attributes" limit mean for logs?

I'll open a tracking issue.

But we don't think it's a blocker for this PR since it's a pre-existing issue for Log attributes.

**Comment by reyang** [comment]
> But we don't think it's a blocker for this PR since it's a pre-existing issue for Log attributes.

+1

File: specification/logs/api.md [hunk 1]
```diff
@@ -185,8 +185,7 @@ formatted as an [event](./data-model.md#events).
 * [Severity Number](./data-model.md#field-severitynumber) (optional)
 * [Severity Text](./data-model.md#field-severitytext) (optional)
 * [Body](./data-model.md#field-body) (optional)
-* [Attributes](./data-model.md#field-attributes) (optional) Event `Attributes` conform
-  to the [standard definition](../common/README.md#standard-attribute) of an attribute.
+* [Attributes](./data-model.md#field-attributes) (optional)
```
Comment (trask) [inline specification/logs/api.md:188]: I wonder if we can limit this to standard attributes plus "maps" and "arrays of maps"

where map values can be
* primitives
* heterogenous array of primitives
* maps
* array of maps

in other words, do we need these?
* non-heterogeneous arrays
* arrays of arrays

Comment (pellared) [reply specification/logs/api.md:188]: I think that such limitation would just add more complexity given that EmitLogRecord needs to support it. Do we have any prototype which does it?

Comment (lmolkova) [reply specification/logs/api.md:188]: Agree with @pellared. If backends provide ergonomics/optimizations for maps, it'll be a good factor for users to use maps. Adding artificial limitation forces to have different attribute types (https://github.com/open-telemetry/opentelemetry-specification/issues/4201) and increases complexity.

But I support these limitations on the semconv side. I'd even consider limiting semconv-supported types to something that can be flattened to a set of standard attributes (i.e. no arrays with complex types).

Comment (trask) [reply specification/logs/api.md:188]: my desire (at least in Java), is precisely to simplify things and avoid the problem that Go is facing in #4201, see https://github.com/open-telemetry/opentelemetry-java/pull/6958

Comment (reyang) [reply specification/logs/api.md:188]: I think we need to have some clarity regarding how maintainers should think about it. For example, C++ has [multimap](https://en.cppreference.com/w/cpp/container/multimap), the developer working on OpenTelemetry C++ and the exporters would need to think:
1. Do I need to support such type in the API.
2. How do I serialize it in OTLP protobuf or JSON.
3. How do I serialize it in other exporters.

Comment (trask) [reply specification/logs/api.md:188]: I think we've gone off topic here. This sounds like an ergonomics question (reusing existing language data types). It's a good question, but I'd suggest someone from C++ SIG open a separate issue about it.

**Comment by felixbarny** [comment]
Is this just about enabling complex attributes for events or also for logs? In order to be able to map structured data according to the [recommendations for logging bridges](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model-appendix.md) into attributes, I think it's required for logs to also support complex attributes.

Example:

```
logger.info('hi', {foo: 'bar', baz: { qux: 'quux'}})
{
  "body": "hi",
  "attributes": {
    "foo": "bar",
    "baz": { qux: "quux"}
  }
}
```

**Comment by trask** [comment]
> Is this just about enabling complex attributes for events or also for logs?

Logs already support complex attributes (and is stable)

**Review by jsuereth — APPROVED** [review]
TL;DR; I'm in for this direction and this change, but I suspect we may need more language around this change.

This diff is much smaller than I anticipated given.

- As others call out, we should update attribute length guards to include these in the future.
- As discussed in the specification meeting, we should provide guidance around how to interact with rich attributes and "flat" attributes.  E.g. is  `"x": { "y": value }` the same as `"x.y": value` ?  (as you know, we've guarded this in semantic conventions)

Approving, under the assumption we all agree on this direction and the follow-on PRs *are* follow-on PRs.

**Comment by trask** [comment]
> As others call out, we should update attribute length guards to include these in the future.

this is tracked in #4335

> As discussed in the specification meeting, we should provide guidance around how to interact with rich attributes and "flat" attributes.  E.g. is  `"x": { "y": value }` the same as `"x.y": value` ?  (as you know, we've guarded this in semantic conventions)

there are a lot of different pieces to this in my mind, I've added it to next week's spec SIG meeting, I'd like to understand which parts you think are required before pre- event stability

**Review by jack-berg — APPROVED** [review]

**Comment by reyang** [comment]
@carlosalberto maybe consider this in 1.41.0 release #4360.

**Review by cijothomas — APPROVED** [review]

**Comment by trask** [comment]
Leaving this PR open for now, but please do not merge, as it will be superseded by #4352 if/once that is merged

**Comment by reyang** [comment]
#4352 is merged, closing this one.
